lib/archive/util.go

```diff
@@ -0,0 +1,118 @@
+	CopyFrom = false
+)
+
+// GenerateFilterSpec will populate the appropriate relative Rebase and Strip paths based on the supplied scenarios. Inclusion/Exclusion should be constructed separately. Please also note that any mount that exists before the copy target that is not primary and comes before the primary target will have a bogus filterspec, since it would not be written or read to.
+func GenerateFilterSpec(copyPath string, mountPoint string, primaryTarget bool, direction bool) FilterSpec {
+	var filter FilterSpec
+
+	// NOTE: this solidifies that this function is very heavily designed for docker cp behavior.
+	// which is why this should mainly only be used by the docker personality
+	// scrub trailing '/' before passing copyPath along
+	copyPath = strings.TrimSuffix(copyPath, "/")
+
+	// Note I know they are just booleans, if that changes then this statement will not need to.
+	if direction == CopyTo {
+		filter = generateCopyToFilterSpec(copyPath, mountPoint, primaryTarget)
+	} else {
+		filter = generateCopyFromFilterSpec(copyPath, mountPoint, primaryTarget)
+	}
+
+	filter.Exclusions = make(map[string]struct{})
+	filter.Inclusions = make(map[string]struct{})
+	return filter
+}
+
+func generateCopyFromFilterSpec(copyPath string, mountPoint string, primaryTarget bool) FilterSpec {
+	var filter FilterSpec
+
+	//we only need the right most element.
+	_, first := filepath.Split(copyPath)
+
+	// primary target was provided so we wil need to split the target and take the right most element for the rebase.
+	// then strip set the strip as the target path.
+	if primaryTarget {
+		filter.RebasePath = removeLeadingSlash(first)
+		filter.StripPath = removeLeadingSlash(strings.TrimPrefix(copyPath, mountPoint))
+		return filter
+	}
+
+	// non primary target was provided. in this case we will need rebase to include the right most member of the target(or "/") joined to the front of the mountPath - the target path. 3
+	filter.RebasePath = removeLeadingSlash(filepath.Join(first, strings.TrimPrefix(mountPoint, copyPath)))
+	filter.StripPath = ""
+	return filter
+}
+
+func generateCopyToFilterSpec(copyPath string, mountPoint string, primaryTarget bool) FilterSpec {
+	var filter FilterSpec
+
+	// primary target was provided so we will need to rebase header assets for this mount to have the target in front for the write.
+	if primaryTarget {
+		filter.RebasePath = removeLeadingSlash(strings.TrimPrefix(copyPath, mountPoint))
+		filter.StripPath = ""
+		return filter
+	}
+
+	// non primary target, this implies that the asset header has part of the mount point path in it. We must strip out that part since the non primary target will be mounted and be looking at the world from it's own root "/"
+	filter.RebasePath = ""
+	filter.StripPath = removeLeadingSlash(strings.TrimPrefix(mountPoint, copyPath))
+
+	return filter
+}
+
+func AddMountInclusionsExclusions(currentMount string, filter *FilterSpec, mounts []string, copyTarget string) error {
```

Comment:
this method should be one method in FilterSpec, and the `mounts []string` should be just `exclusion []string`, which is the term understand by the FilterSpec.
Mount should be excluded, but that's the persona's knowledge